vtuber/source.h

```diff
@@ -1,31 +1,39 @@
 #pragma once
 
+#include "verilog/dtype.h"
 #include <fstream>
+#include <optional>
 #include <string>
 #include <vector>
 
 template <typename T> class Source {
 public:
   Source() {}
   virtual ~Source() = default;
-  virtual ::std::vector<T> get() = 0;
+  virtual ::std::optional<::std::vector<T>> get() = 0;
 };
 
 template <typename T> class VintLineSource : public Source<T> {
-  static_assert(verilog::is_vint_v<T>,
-                "T must be a vint type, not array, struct, or union");
+  static_assert(verilog::is_dtype_v<T>, "T must be a verilog type");
 
 public:
   VintLineSource(const ::std::string &filename_) : Source<T>() {
     filename = filename_;
   }
 
-  ::std::vector<T> get() override {
-    ::std::vector<T> data;
+  ::std::optional<::std::vector<T>> get() override {
     ::std::string line;
     ::std::ifstream inputFile(filename);
```

Comment:
I would suggest istream for supporting more input format

Reply:
Yes I consider that too. Also it is easier to implement test.